api/envoy/config/common/matcher/v3/matcher.proto

```diff
@@ -16,6 +19,93 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
 // [#protodoc-title: Unified Matcher API]
 
+// A MatchTree defines a matching tree that can be used to match incoming data, arriving at a match action that indicates what action to take in response to the match result.
+// This structure allows for traversal to happen as data is becoming available, allowing for optimized matching.
+//
+// We talk about an "associated operation" throughout this docs: this is the context dependent operation that is configured with this match tree. For example, this could be a HTTP/TCP filter.
+message MatchTree {
+  // The action that should be taken upon matching.
+  message MatchAction {
+    oneof action {
+      option (validate.required) = true;
+
+      // The associated operation should be skipped.
+      bool skip = 1 [(validate.rules).bool = {const: true}];
+
+      // The specified callback value should be passed to the associated operation.
+      string callback = 2;
+    }
+  }
+
+  message MatchLeaf {
+    message LeafMatcher {
+      oneof matcher {
+        option (validate.required) = true;
+
+        MatchPredicate predicate = 1;
+
+        // [#not-implemented-hide:]
+        core.v3.TypedExtensionConfig typed_config = 2;
+      }
+
+      // The action to perform if the match evaluates to true.
+      MatchAction action = 3 [(validate.rules).message = {required: true}];
+    }
+
+    // A linear set of match conditions that will be evaluated in order.
+    // Each matcher will be attempted in order until one succeeds.
+    // If none of the matchers match, the no_match_action will be performed if specified.
+    repeated LeafMatcher matchers = 1 [(validate.rules).repeated = {min_items: 1}];
+
+    // Action to perform if none of the above matchers match.
+    MatchAction no_match_action = 3;
+  }
+
+  message SublinearMatcher {
+    message MultimapMatcher {
+      // The key to apply this matcher for.
+      string key = 1 [(validate.rules).string = {min_len: 1}];
```

Comment:
I wonder should this be a message to give more flexibility later, I guess some namespace might have a slight different definition of key, like the dynamic metadata could have multiple keys instead of just 1.